### PR TITLE
[REVIEW] expose degree calculation kernel

### PR DESCRIPTION
**Pull request by Iroy30**

### CODE REVIEW

File: cpp/include/functions.h [hunk 1]
```diff
@@ -144,4 +144,4 @@ gdf_error gdf_delete_transposed_adj_list(gdf_graph *graph);
  * @return GDF_SUCCESS upon successful completion. */
 gdf_error gdf_get_two_hop_neighbors(gdf_graph* graph, gdf_column* first, gdf_column* second);
```
Comment (afender) [inline cpp/include/functions.h]: Missing doc

File: python/cugraph/graph/c_graph.pyx [hunk 2]
```diff
@@ -452,3 +452,33 @@ class Graph:
         err = gdf_add_adj_list(g)
         cudf.bindings.cudf_cpp.check_gdf_error(err)
         return g.adjList.offsets.size - 1
+
+    def in_degree(self):
```
Comment (afender) [inline python/cugraph/graph/c_graph.pyx]: it seems to me that when x=1 it does the out_degree (from destinations indices)

File: python/cugraph/graph/c_graph.pyx [hunk 3]
```diff
@@ -452,3 +452,33 @@ class Graph:
         err = gdf_add_adj_list(g)
         cudf.bindings.cudf_cpp.check_gdf_error(err)
         return g.adjList.offsets.size - 1
+
```
Comment (afender) [inline python/cugraph/graph/c_graph.pyx]: we need doc here as well :)

File: python/cugraph/graph/c_graph.pyx [hunk 4]
```diff
@@ -452,3 +452,33 @@ class Graph:
...
+    def out_degree(self):
+        return self._degree(x=2)
+
+    def degree(self):
+        return self._degree()
+
+    def _degree(self, x = 0):
+        cdef uintptr_t graph = self.graph_ptr
+        cdef gdf_graph* g = < gdf_graph *> graph
+        err = gdf_add_adj_list(g)
```
Comment (afender) [inline python/cugraph/graph/c_graph.pyx]: It is a bit extreme to compute the adj list here just to get the number of vertices.
First, we should check the transposed too.
If both are missing we could do n=max (max(src),max(dest)) which would be much faster

Ideally, when adj list and/or transposed adj list exist we should get the degree from the `offset` which would be much faster. Just need a simple kernel that does `deg[i] = offset[i+1]-offset[i]`

**Comment by afender** [comment]
Nx has the option to return the degree of a single node or of a subset of nodes: https://networkx.github.io/documentation/stable/reference/generated/networkx.classes.function.degree.html#networkx.classes.function.degree

**Comment by afender** [comment]
@BradReesWork we are going to expose 3 python functions (like networkX). They will have native kernels to get the in/out/total degree both from edge list or adj list natively. This is documented in #188.

File: cpp/include/functions.h (hunk 1 above)
Comment (Iroy30) [reply cpp/include/functions.h]: added

File: python/cugraph/graph/c_graph.pyx (hunk 3 above)
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: added

File: python/cugraph/graph/c_graph.pyx (hunk 4 above)
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: If only edge list is present, adjacency list is calculated to get vertex identifiers and their corresponding degree. 
Added kernel for degree calculation from offsets

File: cpp/src/graph_utils.cuh [hunk 5]
```diff
@@ -325,6 +325,15 @@ namespace cugraph
 			if (degree[i] == 0)
 				bookmark[i] = 1.0;
 	}
+
+        template<typename IndexType, typename ValueType>
+        __global__ void __launch_bounds__(CUDA_MAX_KERNEL_THREADS)
+        degree_offsets(const IndexType n, const IndexType e, const IndexType *ind, IndexType *degree) {
+                for (int i = threadIdx.x + blockIdx.x * blockDim.x; i < n; i += gridDim.x * blockDim.x)
+                        atomicAdd(&degree[i], ind[i+1]-ind[i]);
+        }
```
Comment (afender) [inline cpp/src/graph_utils.cuh]: I think we do not need an atomic here. Each thread `i` will write its own output only once at the `i`th location in the degree array

File: python/cugraph/graph/c_graph.pyx [hunk 6]
```diff
@@ -452,3 +452,149 @@ class Graph:
...
+        cdef uintptr_t degree_col_ptr = create_column(degree_col)
+        err = gdf_degree(g, <gdf_column*>degree_col_ptr, <int>x)
+        cudf.bindings.cudf_cpp.check_gdf_error(err)
+
+        if vertex_subset is None:
+            df['vertex'] = vertex_col
+            df['degree'] = degree_col
+        else:
+            df['vertex'] = cudf.Series(np.asarray(vertex_subset, dtype=np.int32))
+            df['degree'] = cudf.Series(np.asarray([degree_col[i] for i in vertex_subset], dtype=np.int32))
```
Comment (afender) [inline python/cugraph/graph/c_graph.pyx]: Do I understand correctly that relevant `degree_col[i]` are copied into a new gpu memory location and that `degree_col` is automatically freed?

Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: yes, it's copied into a new memory location, degree_col and vertex_col are freed

File: cpp/src/graph_utils.cuh (hunk 5 above)
Comment (Iroy30) [reply cpp/src/graph_utils.cuh]: removed atomic add

File: cpp/src/cugraph.cu [hunk 7]
```diff
@@ -239,45 +239,69 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+    dim3 nthreads, nblocks;
+    nthreads.x = min(n, CUDA_MAX_KERNEL_THREADS);
+    nthreads.y = 1;
+    nthreads.z = 1;
+    nblocks.x = min((n + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+    nblocks.y = 1;
+    nblocks.z = 1;
+
+    switch (col_ptr->dtype) {
+      case GDF_INT32:   cugraph::degree_offsets<int, float> <<<nblocks, nthreads>>>(n, e, (int*)col_ptr->data, (int*)degree->data);break;
```
Comment (seunghwak) [inline cpp/src/cugraph.cu]: Better use static_cast<int*>() instead of C style type casting.

File: cpp/src/cugraph.cu [hunk 8]
```diff
@@ -239,45 +239,69 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+    dim3 nthreads, nblocks;
+    nthreads.x = min(e, CUDA_MAX_KERNEL_THREADS);
+    nthreads.y = 1;
+    nthreads.z = 1;
+    nblocks.x = min((e + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+    nblocks.y = 1;
+    nblocks.z = 1;
+    
+    switch (col_ptr->dtype) {
+      case GDF_INT32:   cugraph::degree_coo<int, float> <<<nblocks, nthreads>>>(n, e, (int*)col_ptr->data, (int*)degree->data);break;
```
Comment (seunghwak) [inline cpp/src/cugraph.cu]: Better use static_cast<int*>() instead of C style type casting.

File: cpp/src/cugraph.cu [hunk 9]
```diff
@@ -239,45 +239,69 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+      //case GDF_FLOAT64
+      default: return GDF_UNSUPPORTED_DTYPE;
     }
   }
   return GDF_SUCCESS;
+  }
```
Comment (seunghwak) [inline cpp/src/cugraph.cu]: Check indentation here.

File: cpp/src/cugraph.cu [hunk 10]
```diff
@@ -239,45 +239,69 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+    n = graph->adjList->offsets->size -1;
+    e = graph->adjList->indices->size;
+  }
+  else {
+    n = graph->transposedAdjList->offsets->size - 1; 
+    e = graph->transposedAdjList->indices->size;
+  } 
+
+  if(x!=1) { // out-degree
+  if(graph->adjList)
```
Comment (seunghwak) [inline cpp/src/cugraph.cu]: Check indentation here.

File: cpp/src/cugraph.cu [hunk 11]
```diff
@@ -239,45 +239,69 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+  if(x!=1) { // out-degree
+  if(graph->adjList) 
+    gdf_degree_impl(n, e, graph->adjList->offsets, degree, true);
+  
+  else 
+    gdf_degree_impl(n, e, graph->transposedAdjList->indices, degree, false);
+  }
+
+  if(x!=2) { // in-degree
+  if(graph->adjList)
```
Comment (seunghwak) [inline cpp/src/cugraph.cu]: Check indentation here.

File: python/cugraph/graph/c_graph.pyx [hunk 12]
```diff
@@ -465,19 +465,30 @@ class Graph:
...
         cdef uintptr_t graph = self.graph_ptr
         cdef gdf_graph* g = < gdf_graph *> graph
-        err = gdf_add_adj_list(g)
-        cudf.bindings.cudf_cpp.check_gdf_error(err)
 
-        n = g.adjList.offsets.size - 1
+        cdef gdf_adj_list* adj_ptr
+
+        if(g.adjList):
+            n = g.adjList.offsets.size - 1
```
Comment (seunghwak) [inline python/cugraph/graph/c_graph.pyx]: Better use num_vertices(). We also need to update num_vertices() to check g.transposedAdjList if g.adjList is nullptr.

File: cpp/src/cugraph.cu [hunk 13]
```diff
@@ -239,6 +239,48 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
     return GDF_SUCCESS;
 }
 
+gdf_error  gdf_degree_coo (gdf_graph *graph, gdf_column *degree, int x) {
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/cugraph.cu]: I believe this is an outdated version of the code you  are somehow seeing

File: python/cugraph/graph/c_graph.pyx [hunk 14]
```diff
@@ -452,3 +452,33 @@ class Graph:
         err = gdf_add_adj_list(g)
         cudf.bindings.cudf_cpp.check_gdf_error(err)
         return g.adjList.offsets.size - 1
+
+    def in_degree(self):
+        return self._degree(x=1)
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: I believe this is an outdated version of the code you  are somehow seeing

File: cpp/src/cugraph.cu [hunk 15]
```diff
@@ -239,6 +239,48 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+  nblocks.y = 1;
+  nblocks.z = 1;
+
+  if(x!=1) {
+    gdf_error err = gdf_add_edge_list(graph);
+    if (err != GDF_SUCCESS)
+      return err;
+
+  switch (graph->edgeList->src_indices->dtype) {
+    case GDF_INT32:   cugraph::degree_coo<int, float> <<<nblocks, nthreads>>>(n, e, (int*)graph->edgeList->src_indices->data, (int*)degree->data);break;
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/cugraph.cu]: I believe this is an outdated version of the code you are somehow seeing.

File: cpp/src/cugraph.cu [hunk 16]
```diff
@@ -239,6 +239,48 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+  nblocks.x = min((e + nthreads.x - 1) / nthreads.x, CUDA_MAX_BLOCKS);
+  nblocks.y = 1;
+  nblocks.z = 1;
+
+  if(x!=1) {
+    gdf_error err = gdf_add_edge_list(graph);
+    if (err != GDF_SUCCESS)
+      return err;
+
+  switch (graph->edgeList->src_indices->dtype) {
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/cugraph.cu]: I believe this is an outdated version of the code you  are somehow seeing

File: cpp/src/cugraph.cu [hunk 17]
```diff
@@ -239,6 +239,48 @@ gdf_error gdf_add_transposed_adj_list_impl (gdf_graph *graph) {
...
+    case GDF_INT32:   cugraph::degree_coo<int, float> <<<nblocks, nthreads>>>(n, e, (int*)graph->edgeList->src_indices->data, (int*)degree->data);break;
+    //case GDF_FLOAT64:   degree_coo<int, float> <<<nblocks, nthreads>>>(n, e, src, degree);
+    default: return GDF_UNSUPPORTED_DTYPE;
+    }
+  }
+  if(x!=2) {
+    gdf_error err = gdf_add_edge_list(graph);
+    if (err != GDF_SUCCESS)
+      return err;
+  switch (graph->edgeList->src_indices->dtype) {
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/cugraph.cu]: I believe this is an outdated version of the code you  are somehow seeing

File: cpp/src/cugraph.cu (hunk 15 above)
[the remark this answers is not in the document]
Comment (Iroy30) [reply cpp/src/cugraph.cu]: i think github is having issues. It is weirdly displaying files without latest commit. Anyway, lot of issues are solved in the latest commit. I will fix others

File: python/cugraph/graph/c_graph.pyx (hunk 12 above)
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: modified num_vertices() for checking transposed adj List

File: cpp/src/cugraph.cu (hunk 8 above)
Comment (Iroy30) [reply cpp/src/cugraph.cu]: edited

File: python/cugraph/graph/c_graph.pyx [hunk 18]
```diff
@@ -452,3 +452,33 @@ class Graph:
...
+        err = gdf_add_adj_list(g)
+        cudf.bindings.cudf_cpp.check_gdf_error(err)
+
+        n = g.adjList.offsets.size - 1
+        df = cudf.DataFrame()
+        df['vertex'] = cudf.Series(np.zeros(n, dtype=np.int32))
+        cdef uintptr_t identifier_ptr = create_column(df['vertex'])
+        err = g.adjList.get_vertex_identifiers(<gdf_column*>identifier_ptr)
+        cudf.bindings.cudf_cpp.check_gdf_error(err)
+
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: added deletion of column

File: python/cugraph/graph/c_graph.pyx [hunk 19]
```diff
@@ -452,3 +452,33 @@ class Graph:
...
+        df['vertex'] = cudf.Series(np.zeros(n, dtype=np.int32))
+        cdef uintptr_t identifier_ptr = create_column(df['vertex'])
+        err = g.adjList.get_vertex_identifiers(<gdf_column*>identifier_ptr)
+        cudf.bindings.cudf_cpp.check_gdf_error(err)
+        
+        df['degree'] = cudf.Series(np.zeros(n, dtype=np.int32))
+        cdef uintptr_t degree_col_ptr = create_column(df['degree'])
+        err = gdf_degree_coo(g, <gdf_column*>degree_col_ptr, <int>x)
+        cudf.bindings.cudf_cpp.check_gdf_error(err)
+
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/graph/c_graph.pyx]: added deletion of column

File: python/cugraph/graph/test_graph.py [hunk 20]
```diff
@@ -14,7 +14,7 @@
 import numpy as np
 import pytest
 from scipy.io import mmread
-
+import networkx as nx
```
[the remark this answers is not in the document]
Comment (Iroy30) [reply python/cugraph/graph/test_graph.py]: edited

File: cpp/src/cugraph.cu (hunk 7 above)
Comment (Iroy30) [reply cpp/src/cugraph.cu]: edited

**Pull request merged**